File: Backend/src/nisha/infrastructure/websocket/protocol.py
"""NISHA Binary Frame Protocol implementation."""

import struct
import json
from typing import Dict, Any, Optional
from dataclasses import dataclass

@dataclass
class NISHAFrame:
    """Binary frame protocol for NISHA streams."""
    
    MAGIC = b'NI'
    VERSION = 1
    
    # Stream types
    TELEMETRY = 0x01
    VIDEO = 0x02
    AUDIO = 0x03
    LOCATION = 0x04
    
    # Priorities
    CRITICAL = 1
    EVENT = 2
    CONTINUOUS = 3
    BACKGROUND = 4
    
    magic: bytes
    version: int
    stream_type: int
    priority: int
    sequence: int
    timestamp: int
    payload_len: int
    metadata: Dict[str, Any]
    payload: bytes

    @classmethod
    def from_bytes(cls, data: bytes) -> 'NISHAFrame':
        """Parse raw bytes into a NISHAFrame."""
        if len(data) < 16:
            raise ValueError("Data too short for NISHA header")
            
        offset = 0
        
        # Header (16 bytes)
        magic = data[offset:offset+2]; offset += 2
        if magic != cls.MAGIC:
            raise ValueError(f"Invalid magic: {magic}")
            
        version = data[offset]; offset += 1
        if version != cls.VERSION:
            raise ValueError(f"Unsupported version: {version}")
            
        stream_type = data[offset]; offset += 1
        priority = data[offset]; offset += 1
        # Skip reserved byte
        offset += 1
        
        sequence = struct.unpack('>I', data[offset:offset+4])[0]; offset += 4
        timestamp = struct.unpack('>Q', data[offset:offset+8])[0]; offset += 8
        payload_len = struct.unpack('>I', data[offset:offset+4])[0]; offset += 4
        
        # Metadata (2 bytes length + JSON payload)
        if len(data) < offset + 2:
            raise ValueError("Metadata length missing")
        meta_len = struct.unpack('>H', data[offset:offset+2])[0]; offset += 2
        
        if len(data) < offset + meta_len:
            raise ValueError("Metadata content truncated")
        meta_json = data[offset:offset+meta_len].decode('utf-8')
        metadata = json.loads(meta_json); offset += meta_len
        
        # Payload
        payload = data[offset:offset+payload_len]
        
        return cls(
            magic=magic,
            version=version,
            stream_type=stream_type,
            priority=priority,
            sequence=sequence,
            timestamp=timestamp,
            payload_len=payload_len,
            metadata=metadata,
            payload=payload
        )

    def to_bytes(self) -> bytes:
        """Serialize NISHAFrame to bytes."""
        meta_json = json.dumps(self.metadata).encode('utf-8')
        meta_len = len(meta_json)
        
        header = struct.pack(
            '>2sBBBB I Q I',
            self.MAGIC,
            self.VERSION,
            self.stream_type,
            self.priority,
            0, # reserved
            self.sequence,
            self.timestamp,
            len(self.payload)
        )
        
        return header + struct.pack('>H', meta_len) + meta_json + self.payload

File: mobile-agent/frontend/src/utils/protocol.ts
/**
 * NISHA Frame Protocol implementation for Mobile.
 * Handles binary header construction for streaming packets to the Master node.
 */

export enum StreamType {
  TELEMETRY = 0x01,
  VIDEO = 0x02,
  AUDIO = 0x03,
  LOCATION = 0x04,
  CONTROL = 0xFF,
}

export enum Priority {
  LOW = 0x00,
  HIGH = 0x01,
}

export class NISHAFrame {
  static encode(
    streamType: StreamType,
    priority: Priority,
    sequence: number,
    rssi: number,
    battery: number,
    payload: Uint8Array,
    customMetadata: object = {}
  ): Uint8Array {
    const metadata = { rssi, battery, ...customMetadata };
    const metaJson = JSON.stringify(metadata);
    const metaBytes = new TextEncoder().encode(metaJson);


    const HEADER_SIZE = 24;
    const frameSize = HEADER_SIZE + metaBytes.length + payload.length;
    const buffer = new Uint8Array(frameSize);
    const view = new DataView(buffer.buffer);

    // [0-1] Magic 'NI'
    view.setUint8(0, 0x4E); // 'N'
    view.setUint8(1, 0x49); // 'I'

    // [2] Version
    view.setUint8(2, 0x01);

    // [3] Stream Type
    view.setUint8(3, streamType);

    // [4] Priority
    view.setUint8(4, priority);

    // [5] Reserved
    view.setUint8(5, 0x00);

    // [6-9] Sequence (uint32)
    view.setUint32(6, sequence, false);

    // [10-17] Timestamp (uint64)
    const ts = BigInt(Date.now());
    view.setBigUint64(10, ts, false);

    // [18-21] Payload Len (uint32)
    view.setUint32(18, payload.length, false);

    // [22-23] Meta Len (uint16)
    view.setUint16(22, metaBytes.length, false);

    // Fill metadata starting at index 24 (HEADER_SIZE)
    buffer.set(metaBytes, HEADER_SIZE);

    // Fill payload starting after metadata
    buffer.set(payload, HEADER_SIZE + metaBytes.length);

    return buffer;
  }
}
